Data Modeling and Permissions for a Leaning Management System (#138)
directus-labs/guest-authoring: Developing-an-Interactive-Educational-Platform-with-Directus.md

We can go ahead and save it.

At this point, we will be creating the fields, we will start by creating the name and Description, the fields are the same as the course collection kindly do the same, and if you have forgotten which is normal, you can simply scroll up to revisit it.

The new field is the Content field which is used to hold lesson contents, we will be using the WYSIWYG field as seen below: 

 ![](https://lh7-us.googleusercontent.com/EiGlXJDGfZUaBvT39lxnVG-57qFFag-PnS5OIpfHV7UyHk_wKswiDa55S1BDNjIOxpHPaFGMnkWtk53-CVKvAA6BNzVXPYQUrRArTte4IE9c0j824ubK2E2AsF1t4yKMs1gpF235Ln5htpSMhM17R5Y)

We can save afterwards. For the next field, we will have the Lesson number. For this field we will use the input field and select an integer as all lessons would be whole numbers then we go ahead and save.

![](https://lh7-us.googleusercontent.com/P8UwtU5DYrLjQSY3YX_icJIZ9y-yar58hfPboB0IlUyhhvNTEiSmMfDKBc7szGU-vNRU5op0kmkbL5qcOU8w-BLE4IdGoPDWck0elCmaflUvSfSBgPk_BRaiMo6mtPuuRnpxg_fW-PyL3IwwPaDyrRw)

> Lastly, most lessons are videos, and Directus allows you to manually upload videos or you could opt for a URL, which could be much easier, and for this field, we will use the input option, save afterward. This will be all for our Lesson collection.

Reviewer: Please do this with the other collections definitions - less images, just explain what needs to be made.

The enrollment would be kind of tricky as it deals with linking a few of the other collections, the fields will look like this:

-   Enrollment Date: Date Field 

-   Status: Text or Select Field 

-   Grades/Progress: Text or Numeric Field

We will go ahead to create the enrollment date, status, and grades placing them in their proper fields.

This is what the Collections look like:

![](https://lh7-us.googleusercontent.com/LwtV7Fv9iNLaFrbI2QbXvnXayspC7A7np3txo0PsMSpyG08RzEWPU6msdMuZgxOAHN87HQuI6DH8K8ZpD0G8h20K_T_la5cRoLcvEQe9Xsw3z-VZ-eFHYi57cnLFuAJykvuvUEFuKEXuhZm_Zt-uKwk)

We could give it nice icons to make them look better and interactive. To do this on any collection, click any of your choices and scroll below the fields created, you will find the Icon option, and pick anyone that fits, Directus provides us with a lot of options, we can put in random names and get random icons too as seen below.

![](https://lh7-us.googleusercontent.com/Amg-dwA9bXk2IH0GYgr3ncSC_HnUIQJZRa7GTm-_h5O75w8dNdoJacK1YGDCUYyAxC2J17nThYDWC7PD54Pp1BhMP81KuMyCgICRC9Xws8sBbkugmNyIH30Fl9dWetvz54pTaCKOiFmNvcDwhRd7wb8)

We can go ahead, and add the icons it should look like the image below.

> ![](https://lh7-us.googleusercontent.com/j-0FOXFC2FN0v1U7SdNwIoEv1d3QIzVtnxTpm9a3_WQUbRVN2HuaKyzRMNuBfZ3RkNAVe_pV7LmdqshdakNguHFdCIJosJHi6r1i_p0AvD9gRLWUQtrtP8d33LWU-OPX9gHmne2jBMzPjhQrG608xT0)

Reviewer: Remove

Author: @phazonoverload What exactly am I taking off, the whole section?

Reviewer: Yup. Lines 69 to 79 where the comment was :) It's not important to what you're trying to teach - it's either a visual validation of the steps above, or UI customization

> We can now go ahead to understand and fix the relationship between these collections because they are supposedly dependent on each other. Let's break it down a bit; 
> 
> You do have course, and each course could have multiple lessons, and each lesson could have its quiz, right? We can also say each course should have its instructor.
> 
> As we can see they are heavily connected to give optimum value. Directus has a field for these kinds of relationships, it is called the Relational field. Under this field we have one-to-many, many-to-one, and many-to-many, we need to understand which one fleshes out the described relationship better, so it can be used.
> 
> #### One-to-Many Relationship: 
> 
> In a one-to-many relationship, one record in a collection can be related to multiple records in another collection. This relationship is typically represented by adding a "Many" field in the collection where there are multiple related records.
> 
> For example, in our scenario, the "Courses" collection can have a one-to-many relationship with the "Lessons" collection. This means that one course can have multiple lessons associated with it.
> 
> In Directus, you establish this relationship by adding a "One-to-Many" field in the "Courses" collection that links back to the "Lessons" collection. Each lesson record will then reference the ID of the course it belongs to. Behind the scenes, Directus will set this relationship inside your sequel database and make sure everything plays nice with each other.
> 
> So we will navigate into courses(courses seem to be the mother collection), click the Create field, scroll down to the relational fields, and select the one-to-many:
> 
> ![](https://lh7-us.googleusercontent.com/Fo6OyKHB1BcdnK8fcqNHYdtRsj5Okq_VImsW-pw4ZmipOQjtfZWaWj4B5-WzKY83ldGx_IxPnwGL2bpqhAgvasFBNW3P245knu9JxuAEWI_b75S3ai24sYga-Xtqqm6CXDhqRI0O4nGxP7_ufZw8cXw)
> 
> For our key we will call it "Lessons", the related collection will be Lesson, and the foreign key, which is our key inside the Lesson collection will be course;
> 
> ![](https://lh7-us.googleusercontent.com/vlz0BeZQYgajzhOUO_3srP-lzk6ZhbeU5Ex0NwdhlKAYJH7RtbOciIJSzMnkEMdRCyVhXDVu5MxobZeNpo339d3yPJRPSqPKoJB5hLHpPCC5cpoa9zkm-7fjB5cLNxDJZolEFC_BJ0dGOoAcF54OfY4)
> 
> For our display template, we will use the name "name"
> 
> ![](https://lh7-us.googleusercontent.com/jC-HlWxNsjjZFMYTTekHtok9GjkTXJUAt5czDdWOVWilH7UIQU3_wVv0yItmuJi7hBqIoN4lQzb5n42lO0-LWOEsPTBkRkBFWgssWBPtjBys79VBospZKaMIvS7XKfApVGxoxeu6RVDM1Xm3-iOwDS4)
> 
> Scroll below and enable the item link, then save.
> 
> ![](https://lh7-us.googleusercontent.com/nFNE86v5Qit2Iol0OFq_ORsJvPSlRDY2F_PjrcinuyRtM6pXf_ZjWEnXKwrQJh_9_a0ZVlS9CoxL5ejAdpjU4jxTVGUdELvKA2x1MaaIZTo7pcJd8QIwfGdpvTaCanwNl2wJB7zAwN9jQ0vxrVfUH00)
> 
> Now if we go to our courses we can find lessons as one of the fields;
> 
> ![](https://lh7-us.googleusercontent.com/dKheQgAF8nxtLj8-EG8ddOv9EbkNWAvjNjl3V6jTpaGeF_oTKGxBJOafVSRY2t7WJ_InbrEplvn66PjOtB9qZRJhUMQHmbSyAZ852JxlGXI1y-L_Yksu2r8XMQgy8rcnpwqh4vwgXIYn3ZmjJx82EqA)
> 
> The same will be for lessons too, we will find courses there;
> 
> ![](https://lh7-us.googleusercontent.com/H0wgWBjSElWxVGTbUuTqqTdNYO6D1lD9GuSyhBapvtF84icJbmJQpauO1DOtq9GJ5_H-kzsLW_XcIfDoaZYmZoJ78aHc1zRY5-BFClqR4XO4Tlovg5UWYF7I5rJHEF2yIlBWsnSFqZ0YNrqVIIX4QU0)
> 
> We should do the same for Lesoon and the quiz, each lesson could have its quiz. If you think further you could see another good example of one-to-many, maybe you guessed rightly, yes the quiz should have many questions.
> 
> Going further we could do more of these relationships, like an instructor could take many courses, and a course could have more than one instructor, this is a good example for many to many. We will have to go to the courses or instructors collection, it could be any actually, in our case we wiil go to instructors, just because we  havent used it yet. We will use the many to many relationship, our key will be courses, our related collection will still be courses;
> 
> ![](https://lh7-us.googleusercontent.com/TBYw6vogrDkTFBk6dlPUTT726VnXzpd2hB6-LgNaUCK_yT6EzgdKIf58gvX5a-Sn_mSu93ggqRTAus42z1lfFEM5mUHRmSsuu1Yh1a_uhisu-OjazkAe7pJ8DIywo9864UjQeRWcNSIklyNPGioc-jM)
> 
> If we go back to the data model Directus will do it's magic and viola we have a hidden collection, below;

Reviewer: ```suggestion
## Creating Relationships

The `courses` collection can have a one-to-many relationship with the `lessons` collection. This means that one course can have multiple lessons associated with it.

In the `courses` collection, create a One to Many field called `lessons`. The related collection will be `lessons`, and the foreign key, which is our key inside the related collection will be `course`.

![](https://lh7-us.googleusercontent.com/vlz0BeZQYgajzhOUO_3srP-lzk6ZhbeU5Ex0NwdhlKAYJH7RtbOciIJSzMnkEMdRCyVhXDVu5MxobZeNpo339d3yPJRPSqPKoJB5hLHpPCC5cpoa9zkm-7fjB5cLNxDJZolEFC_BJ0dGOoAcF54OfY4)

To make this easier to select, set the display template to the `name` variable:

![](https://lh7-us.googleusercontent.com/jC-HlWxNsjjZFMYTTekHtok9GjkTXJUAt5czDdWOVWilH7UIQU3_wVv0yItmuJi7hBqIoN4lQzb5n42lO0-LWOEsPTBkRkBFWgssWBPtjBys79VBospZKaMIvS7XKfApVGxoxeu6RVDM1Xm3-iOwDS4)

...[see comment]

If we go back to the data model Directus will have created a junction collection on your behalf:
```

You need to do more to explain the setup with other table relationships. I'd like to see this as a simple bullet list like so: 

> Also create the following One to Many fields:
> - In the `quizzes` collection create a field called `x` with a related collection of `y`

And so on.

![](https://lh7-us.googleusercontent.com/X4LzpJXQn9PDZeUdfYbVp9tyvHrFc0OEJnKJvfeS7BoevGZdpczYmidUcbbxdN0GjhgDyIBQkBRHN57OKxEZr3oqPn8MOCdkYzf2y7PdvkHtDPLzwzK-q5rV-BKwWQPJo2TwEtGmlDP1AH7GWYIVncw)

> If you click on it, we have this;
> 
> ![](https://lh7-us.googleusercontent.com/_yIlUTgXhUC1TcOaJAXqi10CjRcCJOk2sk51bUe5bWDs1G421eAXo2ysK8F-wa4LgaFUZNmf_RJU8zBvEeFOINCTBjoyW-4hiZimLX08UdhfjtoT2m1OeseLYP3eUtx1O5aay3628MKKLwMsyCjXlxA)
> 
> At this point, we are done with our data model, let's navigate to our content;
> 
> ![](https://lh7-us.googleusercontent.com/Y5LREObxI1Dug3eOClw0IqmV8RLXqnbxC_kygIfZR3pgtavzUAHIAupQNMiGoBpdURuPqZ3QBMr7gkeoXuZWOiVMOVZJ8GB_JNQCO0A_3PgU0oLmCCeKcetDQn8CG1BzKsN5-fRD6FF7qZJzDmlwF3I)
> 
> This will take you to the dashboard below;
> 
> ![](https://lh7-us.googleusercontent.com/eCQn-8QraTeRhoF9EHyEomf51uLE0XWTTYDn449Kp8s9GHBtm6eMW80ZeNTYJSFaFalhiC18T0BMG9c00GqzyH1WqXHPRkV3jrNXvKYJh23RFAawjG3-8K6bj2l9kJtWNIkGeApUF4qkVriPO7cziSY)
> -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
> 
> We can go ahead and create our items, by selecting any of the collections above. In our case we will do courses first;
> 
> We will create items in our courses;
> 
> ![](https://lh7-us.googleusercontent.com/xIC1ODpTfX5iSlRi5_QD0BUjYdt2Kf0Hskrz_GTfOKy_Ixi8wEGlJirkJx5U1HKqgv_nsNRGGwjys-IAOpawuiJFlNDZvj-ZpQ5pw192tpHoekr6F-s5JtKPG3B5R71lfO5boAVLhCsP2xhmnQeRsgg)
> 
> If we navigate to: <https://interactive-educational-platform.directus.app/items/Courses>, we will see our raw data for our courses;
> 
> ![](https://lh7-us.googleusercontent.com/0py2EvtrdkF6LaREhqb20F_kuZ0jaukRHzgDj9nHByPfvnX7bpk28q-_m3uFLaBsM1nKS6h5LFS6emXsTdmyaWDJviRthHd45xYekfq2VvQEBBfr9c1AcChVuQEYt_hv_eh3GOtv46k2HSsVgdDBYKM)

Reviewer: ```suggestion
Create 3 courses to test with, and then test accessing it via API by navigating to `https://your-directus-url/items/courses`. You would normally see the same data you authored in the Data Studio as JSON, except you don't. Let's talk about why.
```

You're missing the part where you give public access permissions before this. right now it will give you a 403.

> If we navigate to: <https://interactive-educational-platform.directus.app/items/Lesson>, we will see our raw data for our Lesson; we do have one lesson here, we could create more than one lessons if need be;
> 
> ![](https://lh7-us.googleusercontent.com/c2UkDhLS2PVgBiUm0zSGLgzqRcqXgqoC-RThI4s9QBYp00wW5Jvw6fPZ0w3IyIPycKbUVVRGSea1IqAuKWfPr0vkljU20xHEcpYCCtjZnOXo0GLZuSTE-FtObR7RbeCIoeLw58ZyL1gdPKSASqpRrIo)
> 
> Replace the last words of the links above with the collection name so we can see our raw collections.  Well, the truth is you will not be able to do so, and here is why. You do not have permission to do so, that takes us to our next section, roles, and permissions. Let me break it down a bit, as of now whatever you are seeing is for the public eyes, and directus would not allow you see these contents until the right permissions are given.

Reviewer: ```suggestion
Your request to access data via the Items API provides no authentication details, meaning it is a public request. The public role has no permission to do this.
```

> Roles and Permissions
> ---------------------

Reviewer: ```suggestion
## Roles and Permissions
```

> Roles in Directus can be easily created, navigate to the settings, and select Access Control,
> 
> ![](https://lh7-us.googleusercontent.com/iQrk7gZ_26xp2Q7H2AMPsUVHItEZTEg0cBX3yzh4UY1P2XJwTeTvtH0Hww-l4UOI75FJVpoiOzM8v8X7sGkAqgz7DmTw4BpQdgmVmpCw4ezd0Mq5VoTQYhhaSjz9aRx9lOiIXKfHMM9qFiYNRGndFxc)
> 
> At the top right you will find a plus icon, click it create a role, and save. In our case we will create two roles the Students and the Instructors.
> 
> ![](https://lh7-us.googleusercontent.com/o_JTTKDATPEJ04qM9KEZ_J2EMu2gkOpYwsKBx6CKprmIRTflQNRgChJ5nNjDuPZNZSRfseIy-H3eA500cMPvuPAi6A8YT4I3W2sRlNovXRe4FBY12uv8PTtjQDDybu3lv3VQXb1jrSqiOU-rXhKCF3Y)
> 
> By default we do have the Public and the Admin, we do not need to set permissions for the admin as it has all permissions.
> 
> ![](https://lh7-us.googleusercontent.com/BIdQat51I19KIs3vA3ahnTCxVHJcgqWvTBO9jILTB-k5dgxdghbBrRgboqA5SJB7h7gRWriBrkTCChPzLdlldrsX928dwb2WZzjf1u56KZKAoI9iSRB73VdCAulNXQ1S2subXi7pVMYaDp0-cewq9Bo)

Reviewer: ```suggestion
You can configure roles and their permissions in the Access Control area of the settings module. 
```

> You want to be careful with those you give administrator roles to, as it has no restricted access, well for the public, you want to restrict the public from certain things, like courses, because it most likely going to be a paid course, and even if they weren't, you would only want authenticated users to access the courses, other things we would want to restrict the public from would be everything except the enrollment, we could allow access to contact and chat interfaces if there was. 

Reviewer: ```suggestion
You want to be careful with those you give administrator roles to, as it has access to administer all data, and restrict the public role from certain things, like courses enrollment, because you may only want authenticated users to access them.
```

> ### Permissions
> 
> In permissions, we may want to be precise as to who should see content, create content, make a change to content, and delete content(CRUD). Directus makes the work much easier. We can also apply custom permissions, this will be important in scenarios where we want instructors to be able to create and update courses, remember this is permission given to all instructors, so by doing this, instructors could as well, not just update their courses but that of others, and we do not want that. With custom permissions, we can avoid this.

Reviewer: ```suggestion
Each role has a set of permissions for each collection which describe the ability to see, create, edit, and delete content. Using Directus, you can also apply custom permissions which are more granular than a "yes" or "no", but perhaps only allow instructors to edit their own courses, and students to see those that they are enrolled in.
```

> We will want to go ahead to create permissions for the student, and the instructors.

Reviewer: ```suggestion
Let's create permissions for the students and instructors.
```

You thought we had finished, but not quite right, we do need custom permissions for the students, as we only want students to view published courses, lessons, and quizzes.

We will open up the read permission click custom permissions, and under item permissions, we wiil create a role using a dynamic variable, by clicking the add filter drop dropdown, we are going to go to the quiz questions field, and use the ' Quiz questions to equals $CURRENT_USER.id", and now they would only be able to see the quizzes that belongs to them alone.

![](https://lh7-us.googleusercontent.com/TsilbdWhcx2I1Oy_uNME4cO-BfJMBcHWmyNCZBtj-obb5NGCnfBsrFrSsP8stVrfz12xI2jm1MXIlK4W0ZhX4nk31YJL62EFbie0f59NYZe4K659T3gi5skIkA3FoT5ZN20jtjOJM87MWAgr8-6Q-OI)

Lastly, we only want students to see published courses, or lessons; we navigate to the courses collection, the custom permissions, then item permissions, then we click the add filter dropdown under the rules, select status and let it be equal to published. And viola students can only see published courses, we will go ahead to do same for, lessons, and quiz.

> ![](https://lh7-us.googleusercontent.com/hZBD0NEPySG1pe8Y1hQzGAcf8h-QQJvr9qoEDiphvUaX6p0Mkag2TRs-KRPTOIVMYTlLMBjLZoUejxlv4a5SlIBPwtqLc19qXH943GJs3-bkd8ZbdggqOBi7J8K16seeEv0BLlQuoqJ3WUnqvlOPvbg)

Reviewer: Please put the custom permissions inline, not at the end. For example, you say "Students should be able to view assignments within the courses they are enrolled in." and then show a screenshot where they have "all" access.

> 

Reviewer: I expect to see a section here about actually using this system - you should do this from the student standpoint and the instructor standpoint. Use the APIs to work with them, and mention that they will need to build their application frontend around these API calls and workflows.

Author: @phazonoverload, Thank you for the reviews, but this is not very clear, can this be explained a bit further?

Reviewer: So in this post, you have set up collections, relationships, roles, and permissions. At the end it would be good to actually give a little walk-through on how would be built on top of it to actually build an application. 

I don't want to see a front end actually built, but giving an example of a few API endpoints would be good to manage as both a student and an instructor
